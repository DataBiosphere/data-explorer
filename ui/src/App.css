.app {
  margin: 20px;
}

.legal-disclaimer {
	margin: 20px;
}
