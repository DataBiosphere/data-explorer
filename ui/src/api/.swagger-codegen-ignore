docs/**
git_push.sh
mocha.opts
.travis.yml
