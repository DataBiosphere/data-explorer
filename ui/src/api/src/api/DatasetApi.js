/**
 * Data Explorer Service
 * API Service that reads from Elasticsearch.
 *
 * OpenAPI spec version: 0.0.1
 * 
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 *
 */


import ApiClient from "../ApiClient";
import DatasetResponse from '../model/DatasetResponse';

/**
* Dataset service.
* @module api/DatasetApi
* @version 0.0.1
*/
export default class DatasetApi {

    /**
    * Constructs a new DatasetApi. 
    * @alias module:api/DatasetApi
    * @class
    * @param {module:ApiClient} [apiClient] Optional API client implementation to use,
    * default to {@link module:ApiClient#instance} if unspecified.
    */
    constructor(apiClient) {
        this.apiClient = apiClient || ApiClient.instance;
    }


    /**
     * Callback function to receive the result of the datasetGet operation.
     * @callback module:api/DatasetApi~datasetGetCallback
     * @param {String} error Error message, if any.
     * @param {module:model/DatasetResponse} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Gets dataset information, such as name.
     * @param {module:api/DatasetApi~datasetGetCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/DatasetResponse}
     */
    datasetGet(callback) {
      let postBody = null;


      let pathParams = {
      };
      let queryParams = {
      };
      let headerParams = {
      };
      let formParams = {
      };

      let authNames = [];
      let contentTypes = [];
      let accepts = [];
      let returnType = DatasetResponse;

      return this.apiClient.callApi(
        '/dataset', 'GET',
        pathParams, queryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }


}
