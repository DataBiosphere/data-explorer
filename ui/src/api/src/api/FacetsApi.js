/**
 * Data Explorer Service
 * API Service that reads from Elasticsearch.
 *
 * OpenAPI spec version: 0.0.1
 * 
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 *
 */


import ApiClient from "../ApiClient";
import FacetsResponse from '../model/FacetsResponse';

/**
* Facets service.
* @module api/FacetsApi
* @version 0.0.1
*/
export default class FacetsApi {

    /**
    * Constructs a new FacetsApi. 
    * @alias module:api/FacetsApi
    * @class
    * @param {module:ApiClient} [apiClient] Optional API client implementation to use,
    * default to {@link module:ApiClient#instance} if unspecified.
    */
    constructor(apiClient) {
        this.apiClient = apiClient || ApiClient.instance;
    }


    /**
     * Callback function to receive the result of the facetsGet operation.
     * @callback module:api/FacetsApi~facetsGetCallback
     * @param {String} error Error message, if any.
     * @param {module:model/FacetsResponse} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns facets.
     * @param {Object} opts Optional parameters
     * @param {Array.<String>} opts.filter filter represents selected facet values. Elasticsearch query will be run only over selected facet values. filter is an array of strings, where each string has the format \&quot;esFieldName&#x3D;facetValue\&quot;. Example url /facets?filter&#x3D;Gender&#x3D;female,Region&#x3D;northwest,Region&#x3D;southwest 
     * @param {Array.<String>} opts.extraFacets extraFacets represents the fields selected from the field search box. extraFacets is a list of Elasticsearch field names. In the returned list of facets, the extra facets will come before the facets from ui.json.
     * @param {module:api/FacetsApi~facetsGetCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/FacetsResponse}
     */
    facetsGet(opts, callback) {
      opts = opts || {};
      let postBody = null;


      let pathParams = {
      };
      let queryParams = {
        'filter': this.apiClient.buildCollectionParam(opts['filter'], 'pipes'),
        'extraFacets': this.apiClient.buildCollectionParam(opts['extraFacets'], 'pipes')
      };
      let headerParams = {
      };
      let formParams = {
      };

      let authNames = [];
      let contentTypes = [];
      let accepts = [];
      let returnType = FacetsResponse;

      return this.apiClient.callApi(
        '/facets', 'GET',
        pathParams, queryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }


}
