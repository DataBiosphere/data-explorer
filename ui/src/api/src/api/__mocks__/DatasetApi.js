import DatasetResponse from "../../model/DatasetResponse";

export const mockDatasetGet = jest.fn(callback => {
  let response = new DatasetResponse();
  response.name = "Test Dataset Name";
  callback(null, response, null);
  return response;
});

const mock = jest.fn().mockImplementation(() => {
  return {
    datasetGet: mockDatasetGet
  };
});

export default mock;
