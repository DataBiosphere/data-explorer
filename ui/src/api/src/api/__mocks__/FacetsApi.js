import FacetsResponse from "../../model/FacetsResponse";
import { getFacetsList } from "../../../../../tests/mockutils";

export const mockFacetsGet = jest.fn((request, callback) => {
  let getFacetsResponse = new FacetsResponse();
  getFacetsResponse.count = 210;
  getFacetsResponse.facets = getFacetsList(3, 4);
  callback(null, getFacetsResponse, request);
  return getFacetsResponse;
});

const mock = jest.fn().mockImplementation(() => {
  return {
    facetsGet: mockFacetsGet
  };
});

export default mock;
