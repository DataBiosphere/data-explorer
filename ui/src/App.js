import React, { Component } from "react";
import { MuiThemeProvider, createMuiTheme } from "@material-ui/core/styles";
import Typography from "@material-ui/core/Typography";

import "App.css";
import {
  ApiClient,
  DatasetApi,
  FacetsApi,
  SearchApi
} from "data_explorer_service";
import ExportFab from "components/ExportFab";
import ExportUrlApi from "api/src/api/ExportUrlApi";
import FacetsGrid from "components/facets/FacetsGrid";
import Search from "components/Search";
import Header from "components/Header";

const theme = createMuiTheme({
  typography: {
    fontFamily: ["Lato", "sans-serif"].join(",")
  }
});

const Disclaimer = (
  <Typography style={{ margin: "20px" }}>
    This dataset is publicly available for anyone to use under the terms
    provided by the dataset source (<a href="http://www.internationalgenome.org/data">
      http://www.internationalgenome.org/data
    </a>) and are provided "AS IS" without any warranty, express or implied,
    from Verily Life Sciences, LLC. Verily Life Sciences, LLC disclaims all
    liability for any damages, direct or indirect, resulting from the use of the
    dataset.
  </Typography>
);

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      datasetName: "",
      // What to show in search box by default. If this is the empty string, the
      // react-select default of "Select..." is shown.
      searchPlaceholderText: "",
      // Map from es_field_name to facet data returned from API server /facets call.
      facets: new Map(),
      totalCount: null,
      // Map from es_field_name to a list of selected facet values.
      selectedFacetValues: new Map(),
      // Search results shown in the search drop-down.
      // This is an array of dicts, where each dict has
      // facetName - The name of the facet.
      // facetDescription - The description of the facet.
      // esFieldName - The elasticsearch field name of the facet.
      // facetValue
      searchResults: [],
      // These represent extra facets added via the search box.
      // This is an array of Elasticsearch field names
      extraFacetEsFieldNames: []
    };

    this.apiClient = new ApiClient();
    this.apiClient.basePath = "/api";
    this.facetsApi = new FacetsApi(this.apiClient);
    this.facetsCallback = function(error, data) {
      if (error) {
        console.error(error);
        // TODO(alanhwang): Redirect to an error page
      } else {
        this.setState({
          facets: this.getFacetMap(data.facets),
          totalCount: data.count
        });
      }
    }.bind(this);

    this.searchApi = new SearchApi(this.apiClient);
    this.searchCallback = function(error, data) {
      if (error) {
        console.error(error);
      } else {
        this.setState({
          searchResults: data.search_results.map(searchResult => {
            return {
              facetName: searchResult.facet_name,
              facetDescription: searchResult.facet_description,
              esFieldName: searchResult.elasticsearch_field_name,
              facetValue: searchResult.facet_value
            };
          })
        });
      }
    }.bind(this);

    // Map from facet name to a list of facet values.
    this.filterMap = new Map();
    this.updateFacets = this.updateFacets.bind(this);
    this.handleSearchBoxChange = this.handleSearchBoxChange.bind(this);
  }

  render() {
    if (this.state.facets.size == 0 || this.state.datasetName === "") {
      // Server has not yet responded or returned an error
      return <div />;
    } else {
      return (
        <MuiThemeProvider theme={theme}>
          <div className="app">
            <Header
              datasetName={this.state.datasetName}
              totalCount={this.state.totalCount}
            />
            <Search
              searchPlaceholderText={this.state.searchPlaceholderText}
              searchResults={this.state.searchResults}
              handleSearchBoxChange={this.handleSearchBoxChange}
              selectedFacetValues={this.state.selectedFacetValues}
              facets={this.state.facets}
            />
            <FacetsGrid
              updateFacets={this.updateFacets}
              selectedFacetValues={this.state.selectedFacetValues}
              facets={Array.from(this.state.facets.values())}
            />
            <ExportFab
              exportUrlApi={new ExportUrlApi(this.apiClient)}
              filter={this.filterMapToArray(this.state.selectedFacetValues)}
            />
            {this.state.datasetName == "1000 Genomes" ? Disclaimer : null}
          </div>
        </MuiThemeProvider>
      );
    }
  }

  componentDidMount() {
    this.searchApi.searchGet(this.searchCallback);
    this.facetsApi.facetsGet({}, this.facetsCallback);

    // Call /api/dataset
    let datasetApi = new DatasetApi(this.apiClient);
    let datasetCallback = function(error, data) {
      if (error) {
        // TODO: Show error in snackbar.
        console.error(error);
      } else {
        this.setState({
          datasetName: data.name,
          searchPlaceholderText: data.search_placeholder_text
        });
      }
    }.bind(this);
    datasetApi.datasetGet(datasetCallback);
  }

  getFacetMap(facets) {
    var facetMap = new Map();
    facets.forEach(function(facet) {
      facetMap.set(facet.es_field_name, facet);
    });
    return facetMap;
  }

  handleSearchBoxChange(selectedOptions, action) {
    if (action.action == "clear") {
      // x on right of search box was clicked.
      this.setState({ selectedFacetValues: new Map() });
      this.facetsApi.facetsGet(
        {
          filter: this.filterMapToArray(new Map()),
          extraFacets: this.state.extraFacetEsFieldNames
        },
        this.facetsCallback
      );
    } else if (action.action == "remove-value") {
      // chip x was clicked.
      let parts = action.removedValue.value.split("=");
      this.updateFacets(parts[0], parts[1], false);
    } else if (action.action == "select-option") {
      // Drop-down row was clicked.
      let newExtraFacetEsFieldNames = this.state.extraFacetEsFieldNames;
      newExtraFacetEsFieldNames.push(action.option.esFieldName);
      this.setState({ extraFacetEsFieldNames: newExtraFacetEsFieldNames });
      this.facetsApi.facetsGet(
        {
          filter: this.filterMapToArray(this.state.selectedFacetValues),
          extraFacets: newExtraFacetEsFieldNames
        },
        this.facetsCallback
      );
    }
  }

  /**
   * Updates the selection for a single facet value and refreshes the facets data from the server.
   * */
  updateFacets(esFieldName, facetValue, isSelected) {
    let currentFilterMap = this.state.selectedFacetValues;
    let currentFacetValues = currentFilterMap.get(esFieldName);
    if (isSelected) {
      // Add facetValue to the list of filters for facetName
      if (currentFacetValues === undefined) {
        currentFilterMap.set(esFieldName, [facetValue]);
      } else {
        currentFacetValues.push(facetValue);
      }
    } else if (currentFilterMap.get(esFieldName) !== undefined) {
      // Remove facetValue from the list of filters for facetName
      currentFilterMap.set(
        esFieldName,
        this.removeFacetValue(currentFacetValues, facetValue)
      );
    }
    // Update the state
    this.setState({ selectedFacetValues: currentFilterMap });

    // Update the facets grid.
    this.facetsApi.facetsGet(
      {
        filter: this.filterMapToArray(this.state.selectedFacetValues),
        extraFacets: this.state.extraFacetEsFieldNames
      },
      this.facetsCallback
    );
  }

  // Remove the given facet value from a list of facet values.
  removeFacetValue(valueList, facetValue) {
    let newValueList = [];
    for (let i = 0; i < valueList.length; i++) {
      if (valueList[i] !== facetValue) {
        newValueList.push(valueList[i]);
      }
    }
    return newValueList;
  }

  /**
   * Converts a Map of filters to an Array of filter strings interpretable by
   * the backend
   * @param filterMap Map of esFieldName:[facetValues] pairs
   * @return [string] Array of "esFieldName=facetValue" strings
   */
  filterMapToArray(filterMap) {
    let filterArray = [];
    filterMap.forEach((values, key) => {
      // Scenario where there are no values for a key: A single value is
      // checked for a facet. The value is unchecked. The facet name will
      // still be a key in filterMap, but there will be no values.
      if (values.length > 0) {
        for (let value of values) {
          filterArray.push(key + "=" + value);
        }
      }
    });
    return filterArray;
  }
}

export default App;
