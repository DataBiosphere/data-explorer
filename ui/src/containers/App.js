import './App.css';
import { ApiClient, FacetsApi } from 'data_explorer_service';
import Facet from "../api/src/model/Facet";
import FacetValue from "../api/src/model/FacetValue";
import FacetsResponse from "../api/src/model/FacetsResponse";
import FacetsGrid from "../components/FacetsGrid";
import Header from "../components/Header";

import React, { Component } from 'react';
import { MuiThemeProvider } from "material-ui";

class App extends Component {

  constructor(props) {
    super(props);
      this.state = {
          facetResponse: null
      };
  }

  render() {
    if (this.state.facetResponse == null) {
      // Server has not yet responded or returned an error
      return <div></div>;
    } else {
        return (
            <MuiThemeProvider>
                <div className="app">
                    <Header count = {this.state.facetResponse.count} />
                    <FacetsGrid facetResponse = {this.state.facetResponse} />
                </div>
            </MuiThemeProvider>
        );
    }
  }

  componentDidMount() {
    let apiClient = new ApiClient();
    apiClient.basePath = '/api';
    let api = new FacetsApi(apiClient);
    let callback = function(error, data) {
      if (error) {
        console.error(error);
        // TODO(alanhwang): Redirect to an error page
      } else {
        this.setState({facetResponse: data});
      }
    }.bind(this);
    api.facetsGet(callback);
  }
}

export default App;
