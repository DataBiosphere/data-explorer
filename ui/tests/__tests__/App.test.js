import React from "react";
import App from "../../src/App";
import { mockFacetsGet } from "../../src/api/src/api/FacetsApi";
import { mockDatasetGet } from "../../src/api/src/api/DatasetApi";

jest.mock("../../src/api/src/api/DatasetApi");
jest.mock("../../src/api/src/api/FacetsApi");
("use strict");

beforeEach(() => {
  mockFacetsGet.mockClear();
  mockDatasetGet.mockClear();
});

test("Renders an empty div before receiving all data", () => {
  mockDatasetGet.mockImplementationOnce(() => {});
  const tree = shallow(<App />);

  expect(tree).toMatchSnapshot();
});

test("Renders correctly after receiving all data", () => {
  const tree = shallow(<App />);

  expect(tree).toMatchSnapshot();
});

test("Stringifies queries when calling the API to update Facets", () => {
  const tree = shallow(<App />);
  tree.instance().updateFacets("Facet 1", "FacetValue 1", true);
  tree.instance().updateFacets("Facet 1", "FacetValue 2", true);
  tree.instance().updateFacets("Facet 1", "FacetValue 1", false);

  expect(mockFacetsGet).toHaveBeenCalledWith(
    {},
    tree.instance().facetsCallback
  );
  expect(mockFacetsGet).toHaveBeenCalledWith(
    { filter: ["Facet 1=FacetValue 1"] },
    tree.instance().facetsCallback
  );
  expect(mockFacetsGet).toHaveBeenCalledWith(
    { filter: ["Facet 1=FacetValue 1", "Facet 1=FacetValue 2"] },
    tree.instance().facetsCallback
  );
  expect(mockFacetsGet).toHaveBeenCalledWith(
    { filter: ["Facet 1=FacetValue 2"] },
    tree.instance().facetsCallback
  );
});
