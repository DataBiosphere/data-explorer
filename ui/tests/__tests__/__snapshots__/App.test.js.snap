// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the app after receiving data 1`] = `
<MuiThemeProvider>
  <div
    className="app"
  >
    <Header
      datasetName="Test Dataset Name"
      totalCount={210}
    />
    <FacetsGrid
      facets={
        Array [
          Facet {
            "name": "Facet 1",
            "values": Array [
              FacetValue {
                "count": 10,
                "name": "FacetValue 1",
              },
              FacetValue {
                "count": 20,
                "name": "FacetValue 2",
              },
              FacetValue {
                "count": 30,
                "name": "FacetValue 3",
              },
              FacetValue {
                "count": 40,
                "name": "FacetValue 4",
              },
            ],
          },
          Facet {
            "name": "Facet 2",
            "values": Array [
              FacetValue {
                "count": 10,
                "name": "FacetValue 1",
              },
              FacetValue {
                "count": 20,
                "name": "FacetValue 2",
              },
              FacetValue {
                "count": 30,
                "name": "FacetValue 3",
              },
              FacetValue {
                "count": 40,
                "name": "FacetValue 4",
              },
            ],
          },
          Facet {
            "name": "Facet 3",
            "values": Array [
              FacetValue {
                "count": 10,
                "name": "FacetValue 1",
              },
              FacetValue {
                "count": 20,
                "name": "FacetValue 2",
              },
              FacetValue {
                "count": 30,
                "name": "FacetValue 3",
              },
              FacetValue {
                "count": 40,
                "name": "FacetValue 4",
              },
            ],
          },
        ]
      }
      updateFacets={[Function]}
    />
    <ExportFab />
  </div>
</MuiThemeProvider>
`;

exports[`Renders the app empty before receiving data 1`] = `
<MuiThemeProvider>
  <div
    className="app"
  >
    <Header
      datasetName="Test Dataset Name"
      totalCount={210}
    />
    <FacetsGrid
      facets={
        Array [
          Facet {
            "name": "Facet 1",
            "values": Array [
              FacetValue {
                "count": 10,
                "name": "FacetValue 1",
              },
              FacetValue {
                "count": 20,
                "name": "FacetValue 2",
              },
              FacetValue {
                "count": 30,
                "name": "FacetValue 3",
              },
              FacetValue {
                "count": 40,
                "name": "FacetValue 4",
              },
            ],
          },
          Facet {
            "name": "Facet 2",
            "values": Array [
              FacetValue {
                "count": 10,
                "name": "FacetValue 1",
              },
              FacetValue {
                "count": 20,
                "name": "FacetValue 2",
              },
              FacetValue {
                "count": 30,
                "name": "FacetValue 3",
              },
              FacetValue {
                "count": 40,
                "name": "FacetValue 4",
              },
            ],
          },
          Facet {
            "name": "Facet 3",
            "values": Array [
              FacetValue {
                "count": 10,
                "name": "FacetValue 1",
              },
              FacetValue {
                "count": 20,
                "name": "FacetValue 2",
              },
              FacetValue {
                "count": 30,
                "name": "FacetValue 3",
              },
              FacetValue {
                "count": 40,
                "name": "FacetValue 4",
              },
            ],
          },
        ]
      }
      updateFacets={[Function]}
    />
    <ExportFab />
  </div>
</MuiThemeProvider>
`;
